src/server/controlchan/commands/pasv.rs

```diff
@@ -171,8 +170,36 @@ where
     async fn handle(&self, args: CommandContext<S, U>) -> Result<Reply, ControlChanError> {
         let sender: Option<ProxyLoopSender<S, U>> = args.proxyloop_msg_tx.clone();
         match sender {
-            Some(tx) => self.handle_proxy_mode(args, tx.clone()).await,
+            Some(tx) => self.handle_proxy_mode(args, tx).await,
             None => self.handle_nonproxy_mode(args).await,
         }
     }
 }
+
+pub async fn make_pasv_reply(passive_host: PassiveHost, conn_ip: &Ipv4Addr, port: u16) -> Reply {
+    let p1 = port >> 8;
+    let p2 = port - (p1 * 256);
+    let octets = match passive_host {
+        PassiveHost::IP(ip) => ip.octets(),
+        PassiveHost::FromConnection => conn_ip.octets(),
+        PassiveHost::DNS(ref dns_name) => {
+            let x = dns_name.split(':').take(1).map(|s| format!("{}:2121", s)).next().unwrap();
+            let res_iter = tokio::net::lookup_host(x).await;
+            if res_iter.is_err() {
+                return Reply::new_with_string(ReplyCode::CantOpenDataConnection, format!("Could not resolve DNS address '{}'", dns_name));
+            }
+            let mut iter = res_iter.unwrap();
+            loop {
+                match iter.next() {
+                    None => return Reply::new_with_string(ReplyCode::CantOpenDataConnection, format!("Could not resolve DNS address '{}'", dns_name)),
+                    Some(SocketAddr::V4(ip)) => break ip.ip().octets(),
+                    Some(SocketAddr::V6(_)) => continue,
+                }
+            }
+        }
```

Comment:
Why check `if res_iter.is_err()` and then `res_iter.unwrap()` instead of a `match`?

Reply:
To save on indenting but I don't mind either way.

Reply:
I would prefer match.

Reply:
Matched it